src/Runner.Worker/ActionManager.cs

```diff
@@ -53,30 +53,63 @@ public sealed class ActionManager : RunnerService, IActionManager
         public Dictionary<Guid, ContainerInfo> CachedActionContainers => _cachedActionContainers;
         public async Task<PrepareResult> PrepareActionsAsync(IExecutionContext executionContext, IEnumerable<Pipelines.JobStep> steps)
         {
+            // Assert inputs
             ArgUtil.NotNull(executionContext, nameof(executionContext));
             ArgUtil.NotNull(steps, nameof(steps));
-
-            executionContext.Output("Prepare all required actions");
-            Dictionary<string, List<Guid>> imagesToPull = new Dictionary<string, List<Guid>>(StringComparer.OrdinalIgnoreCase);
-            Dictionary<string, List<Guid>> imagesToBuild = new Dictionary<string, List<Guid>>(StringComparer.OrdinalIgnoreCase);
-            Dictionary<string, ActionContainer> imagesToBuildInfo = new Dictionary<string, ActionContainer>(StringComparer.OrdinalIgnoreCase);
-            List<JobExtensionRunner> containerSetupSteps = new List<JobExtensionRunner>();
-            Dictionary<Guid, IActionRunner> preStepTracker = new Dictionary<Guid, IActionRunner>();
+            var state = new PrepareActionsState
+            {
+                ImagesToBuild = new Dictionary<string, List<Guid>>(StringComparer.OrdinalIgnoreCase),
+                ImagesToPull = new Dictionary<string, List<Guid>>(StringComparer.OrdinalIgnoreCase),
+                ImagesToBuildInfo = new Dictionary<string, ActionContainer>(StringComparer.OrdinalIgnoreCase),
+                PreStepTracker = new Dictionary<Guid, IActionRunner>()
+            };
+            var containerSetupSteps = new List<JobExtensionRunner>();
+            IOUtil.DeleteDirectory(HostContext.GetDirectory(WellKnownDirectory.Actions), executionContext.CancellationToken);
             IEnumerable<Pipelines.ActionStep> actions = steps.OfType<Pipelines.ActionStep>();
-
-            // TODO: Deprecate the PREVIEW_ACTION_TOKEN
-            // Log even if we aren't using it to ensure users know.
-            if (!string.IsNullOrEmpty(executionContext.Global.Variables.Get("PREVIEW_ACTION_TOKEN")))
+            executionContext.Output("Prepare all required actions");
+            var result = await PrepareActionsRecursiveAsync(executionContext, state, actions, 0);
```

Comment:
super nit, but interesting imo: if you change the name to `PrepareActionsAsync` then you could name the depth param

```
var result = await PrepareActionsAsync(executionContext, state, actions, depth: 0);
```